sql: cleanly handle descriptor deletions during schema changes

vivekmenezes opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13166)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: make sync/async schema execution paths somewhat similar

The main difference is that the sync path is capable of retrying
the schema change and report some errors in a response.

Handle sqlbase.ErrDescriptorNotFound in the synchronous code path
and fix the bug below.

fixes #13014

tamird commented:
This first commit message leaves much to be desired; what is the higher level change here? Repeating the code change itself in the commit message is not useful.

---

Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/schema_changer.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbQfr3GTpWdKDuUBLcm:-KbQfr3GTpWdKDuUBLcn:b6euoln) ([raw file](https://github.com/cockroachdb/cockroach/blob/92f1f4d8af166c0813999a9bd56c1cdc991aaaa6/pkg/sql/schema_changer.go#L307)):*
> ```Go
> 			log.Warning(context.TODO(), err)
> 		}
> 	}(&lease)
> ```

what is the point of this argument?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbQg0mcNnK9Vb7o4hkL:b-94b75b)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/session.go:
```diff
@@ -476,17 +477,18 @@ func (scc *schemaChangerCollection) execSchemaChanges(
+					// schema change completion.
+					log.Warning(ctx, err)
+					break
+				}
```
I'm looking at removing this from this PR. Still figuring it out.

vivekmenezes commented:
Hope the commit message is more useful now

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbRUiLMN6ASibrRkAit:-KbRUiLNBKe2PGq8D1IL:b-jhti6w) (no related file):*
This PR is ready for another review. Thanks!

---

*[pkg/sql/schema_changer.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbQfr3GTpWdKDuUBLcm:-KbRU_93IuiR9Mfi_I1a:b5arviy) ([raw file](https://github.com/cockroachdb/cockroach/blob/92f1f4d8af166c0813999a9bd56c1cdc991aaaa6/pkg/sql/schema_changer.go#L307)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is the point of this argument?
</blockquote></details>

removed this argument

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbRUkfuXQwx-A_YCT2b:b-ebvzxn)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Any thoughts on this PR?

tamird commented:
This PR is still not passing tests.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbjvvUbZbf1ae2EsQ1x:b878u0m)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
passing now. Thanks

On Mon, Jan 30, 2017 at 10:13 AM Tamir Duberstein <notifications@github.com>
wrote:

> This PR is still not passing tests.
> ------------------------------
>
> Review status: 0 of 3 files reviewed at latest revision, 2 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbjvvUbZbf1ae2EsQ1x:b878u0m>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/13166#issuecomment-276088053>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBKByOufSb2YrSG5oAvR_e3oZ4SPZks5rXf4XgaJpZM4Lu2xc>
> .
>

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkckIcC-vBlFFygI0l:-KbkckIcC-vBlFFygI0m:b8s36d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L201)):*
> ```Go
> }
> 
> func (sc SchemaChanger) maybeAddDropRename(
> ```

please add a comment to this method, also documenting what it returns

---

*[pkg/sql/schema_changer.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkl3UxDIvRreGCoi6A:-Kbkg_MOqFG0jlTOlFzd:b1jfi76) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L285)):*
> ```Go
> func (sc SchemaChanger) exec() error {
> 	// Acquire lease.
> 	lease, err := sc.AcquireLease()
> ```

If we get a "descriptor missing" error, should we check that we weren't trying to also delete this table, and in that case swallow the error?

---

*[pkg/sql/schema_changer_test.go, line 539 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkjA-DCnz270WIEonH:-KbkjA-EjLXph1hIJnTk:bmy3unv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L539)):*
> ```Go
> 					if atomic.AddInt64(&backfillCount, 1) > 1 {
> 						notifyBackfill()
> 						return context.DeadlineExceeded
> ```

please add a comment here explaining that the DeadlineExceeded is supposed to cause something to retry the backfill

---

*[pkg/sql/schema_changer_test.go, line 680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkjn4eGK749sMKuSdu:-Kbkjn4eGK749sMKuSdv:b-28xijp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L680)):*
> ```Go
> 	<-notification
> 
> 	if _, err := sqlDB.Exec("DROP TABLE t.test"); err != nil {
> ```

so the test used to have a bug because this `DROP` was not in fact executed concurrently with the `CREATE INDEX`? It was usually executed after? I don't understand something from your explanation: you said "the table deletion queued up behind the
schema change waiting to acquire the schema change lease on the table" - where's that queuing happening? There's no such thing in the sync schema change execution path, is there? Shouldn't this DROP have returned an internal error, based on `errExistingSchemaChangeLease`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbklKSxWlRa_uv3lV9X:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
> 8be401d: sql: removed check if a schema change is complete prior to execution
> 
> This brings the synchronous schema change execution code path in line
> with the asynchronous execution path that doesn't call IsDone().

What do you mean by this? I don't see any significant code being deleted here - are the two paths still duplicated?

> The schema change queuing mechanism has been changed to ignore duplicates.
> A duplicate DROP TABLE schema change execution would return
> ErrDescriptorNotFound

Is this tested? Doesn't seem to be.

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KbkkM5UkqTGJ4VDOFTW:b7n50mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
> ```Go
> 
> // Execute the entire schema change in steps.
> func (sc SchemaChanger) exec() error {
> ```

does this refactor have anything to do with the removal of IsDone? it seems unrelated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbklCRRBzpDTg1lhqUl:b-8esq6g)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer_test.go, line 537 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbklgs-imb0bbCMyGNO:-Kbklgs-imb0bbCMyGNP:b-9y92nc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L537)):*
> ```Go
> 				}
> 				if partialBackfill {
> 					if atomic.AddInt64(&backfillCount, 1) > 1 {
> ```

what is the point of this integer? this function is already a no-op the second time it is called.

---

*[pkg/sql/schema_changer_test.go, line 672 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbklslpqGcrcPsZXl7v:-KbklslpqGcrcPsZXl7w:bj817zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L672)):*
> ```Go
> 			"CREATE UNIQUE INDEX bar ON t.test (v)",
> 		); !testutils.IsError(err, sqlbase.ErrDescriptorNotFound.Error()) {
> 			t.Error(err)
> ```

this should be `t.Errorf("unexpected error: %v", err)` since `errr` can be `nil` and that would produce a confusing result.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
The synchronous code path would call IsDone() while the asynchronous code path wouldn't. We've deleted IsDone from the synchronous code path to bring both paths more in line.

ignoring duplicate schema changes is tested in a number of testdata tests. I've added a comment to one of them to make it explicit that that test is testing it. It will fail without the code to ignore the duplicate schema change. I suppose I should add here that IsDone() woud returning true when it saw duplicate schema changes and the prior code would ignore duplicate using IsDone()

---

Review status: 1 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/schema_changer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkckIcC-vBlFFygI0l:-KbkmvJekZEhaXgFnLhF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L201)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment to this method, also documenting what it returns
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-Kbkn0_makVtQQEHEOqI:b-k5owwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this refactor have anything to do with the removal of IsDone? it seems unrelated.
</blockquote></details>

it's unrelated

---

*[pkg/sql/schema_changer.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkl3UxDIvRreGCoi6A:-KbknCHDSOb0gZo2ivJV:bjitsff) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L285)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If we get a "descriptor missing" error, should we check that we weren't trying to also delete this table, and in that case swallow the error?
</blockquote></details>

If we get a descriptor missing error we will return it. We do not swallow the error.

---

*[pkg/sql/schema_changer_test.go, line 537 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbklgs-imb0bbCMyGNO:-Kbl12dBAqJ8qTtnqOkz:bkfc3q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L537)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is the point of this integer? this function is already a no-op the second time it is called.
</blockquote></details>

Changed to a bool, but I admit atomic.Value isn't great for the boolean value case

---

*[pkg/sql/schema_changer_test.go, line 539 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkjA-DCnz270WIEonH:-Kbknz46N-vErwxR9zP6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L539)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment here explaining that the DeadlineExceeded is supposed to cause something to retry the backfill
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 672 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbklslpqGcrcPsZXl7v:-Kbko7nv_yDP6JeqYSJ0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L672)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this should be `t.Errorf("unexpected error: %v", err)` since `errr` can be `nil` and that would produce a confusing result.
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkjn4eGK749sMKuSdu:-KbkoF5X_Sm-zo0AXa_p:b-u8rwu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L680)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so the test used to have a bug because this `DROP` was not in fact executed concurrently with the `CREATE INDEX`? It was usually executed after? I don't understand something from your explanation: you said "the table deletion queued up behind the
schema change waiting to acquire the schema change lease on the table" - where's that queuing happening? There's no such thing in the sync schema change execution path, is there? Shouldn't this DROP have returned an internal error, based on `errExistingSchemaChangeLease`?
</blockquote></details>

errExistingSchemaChangeLease results in the schema change being retried, so the DROP would keep getting retried behind the CREATE INDEX. The bug is that the CREATE INDEX would sometimes fail with a retryable error thereby giving up the lease and letting the DROP move forward execute properly and an error returning with the CREATE INDEX. We now let the CREATE INDEX succeed partially and then test that the DROP deletes the entries created by the CREATE INDEX, and only release the CREATE INDEX once the DROP has succeeded.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-Kbkp58f5Afsva_6fnTA:btes6we)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 4 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/schema_changer.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkl3UxDIvRreGCoi6A:-Kbl2kOC0e2XEm0J7bZG:b5wrfqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L285)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

If we get a descriptor missing error we will return it. We do not swallow the error.
</blockquote></details>

I see that we don't swallow it, I'm asking if we should. Since it's expected for multiple async schema changers to get this error when executing an async DROP, right?

---

*[pkg/sql/schema_changer_test.go, line 680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkjn4eGK749sMKuSdu:-Kbl27zoyxd6fnipyW8z:bkbs1a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L680)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

errExistingSchemaChangeLease results in the schema change being retried, so the DROP would keep getting retried behind the CREATE INDEX. The bug is that the CREATE INDEX would sometimes fail with a retryable error thereby giving up the lease and letting the DROP move forward execute properly and an error returning with the CREATE INDEX. We now let the CREATE INDEX succeed partially and then test that the DROP deletes the entries created by the CREATE INDEX, and only release the CREATE INDEX once the DROP has succeeded.
</blockquote></details>

but the sync path doesn't seem to retry on `errExistingLease`. It uses `isSchemaChangeRetryError()`, which doesn't handle that error. In fact, looks like error handling could be done more uniformly between the sync/async paths (async doesn't use `isSchemaChangeRetryError` and duplicates handling of some errors.
What am I missing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-Kbl6TKK2v1leoHzLZGS:b-rsee6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

it's unrelated
</blockquote></details>

Can you revert it? You're just churning code here, muddying the `git blame` and making the review more difficult. I'm not seeing the upshot.

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-Kbl6MpcCqg9CoPmrh5u:bwxh7cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
> ```
> # This drops the table and the reference to the same table,
> # and tests that a duplicate schema change is only applied once.
> # If the DROP TABLE were applied twice the duplicate would return
> ```

what? where's the duplicate? I don't understand this comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KblIHtWo04RzjkHHMnR:b7s4qlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Can you revert it? You're just churning code here, muddying the `git blame` and making the review more difficult. I'm not seeing the upshot.
</blockquote></details>

I've added another comment to the commit so that it shows up in git blame.

The code for add/drop/rename a table has been taken out of
    SchemaChanger.exec() and placed in maybeAddDropRename()

---

*[pkg/sql/schema_changer.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkl3UxDIvRreGCoi6A:-KblEGWL7vN_xJA3VT6z:bsk8b22) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L285)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I see that we don't swallow it, I'm asking if we should. Since it's expected for multiple async schema changers to get this error when executing an async DROP, right?
</blockquote></details>

There are a number of places where we can get the same error. I prefer not swallowing it.

---

*[pkg/sql/schema_changer_test.go, line 680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkjn4eGK749sMKuSdu:-KblJ4qQWMoL5rcSpGaq:bqkc3ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L680)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but the sync path doesn't seem to retry on `errExistingLease`. It uses `isSchemaChangeRetryError()`, which doesn't handle that error. In fact, looks like error handling could be done more uniformly between the sync/async paths (async doesn't use `isSchemaChangeRetryError` and duplicates handling of some errors.
What am I missing?
</blockquote></details>

`isSchemaChangeRetryError()` returns true on seeing `errExistingLease` . The async path just retries on another node or retries on the same node after it sees the outstanding schema change show up again. 

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-KblOH0Ik6HGiJNoYlim:b-va6s68) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what? where's the duplicate? I don't understand this comment.
</blockquote></details>

Updated the comment. Hope it makes sense now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/schema_changer_test.go, line 680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbkjn4eGK749sMKuSdu:-KblP02sqYApYu0Wefbp:b-3hbufw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae59898477e7bf7dce89eb1af29e60341f0872f/pkg/sql/schema_changer_test.go#L680)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

`isSchemaChangeRetryError()` returns true on seeing `errExistingLease` . The async path just retries on another node or retries on the same node after it sees the outstanding schema change show up again. 
</blockquote></details>

oh I missed that `isSchemaChangeRetryError` considers *everything* to be retryable :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KblRJ2DCch175KaHX6c:b8md9ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I've added another comment to the commit so that it shows up in git blame.

The code for add/drop/rename a table has been taken out of
    SchemaChanger.exec() and placed in maybeAddDropRename()
</blockquote></details>

I'm not sure we're understanding each other; my objection to random changes like this trampling the previous `git blame` isn't about the quality of the commit message attached to the change - rather, it is about obscuring the history of why this code is written the way it is.

The next person to wonder about this code will run git blame and see some random commit with unrelated refactoring, rather than the original change that perhaps changed behaviour, along with that change's rationale.

I'm not saying we should never refactor without behaviour change, but in this instance, I don't see the value, and worse: this refactoring is piggy-backing on some other change, which is really not necessary. If you want to submit a refactor, that's fine, but let it be reviewed separately, please.

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-KblS1-BSWRAPo4XVW2q:b8y52kw) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Updated the comment. Hope it makes sense now
</blockquote></details>

I don't understand it still, but perhaps that's on me.

Separately, it doesn't seem appropriate to talk about schema change mechanics in the `fk` test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KblWuYr1soul4_WKDSA:b-is0941) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I'm not sure we're understanding each other; my objection to random changes like this trampling the previous `git blame` isn't about the quality of the commit message attached to the change - rather, it is about obscuring the history of why this code is written the way it is.

The next person to wonder about this code will run git blame and see some random commit with unrelated refactoring, rather than the original change that perhaps changed behaviour, along with that change's rationale.

I'm not saying we should never refactor without behaviour change, but in this instance, I don't see the value, and worse: this refactoring is piggy-backing on some other change, which is really not necessary. If you want to submit a refactor, that's fine, but let it be reviewed separately, please.
</blockquote></details>

I've send out another commit with that change

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-KblX89_UDyaMF5JfzzH:b-j2zd5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't understand it still, but perhaps that's on me.

Separately, it doesn't seem appropriate to talk about schema change mechanics in the `fk` test.
</blockquote></details>

I'm actually cool with not including this comment. Basically the code I've added in this commit is needed to not break these kinds of tests, but if someone later removed that code (the code that ignores duplicates that I've added here) and it doesn't break any tests (in the future), thats really not a breaking change (its fundamentally not needed). So perhaps we should just not add this comment!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KblYe4KNQ5KuqtGu8ap:b-xxcdwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I've send out another commit with that change
</blockquote></details>

Where?

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-KblYnp-La3uSRoI3ZmC:bc491pk) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I'm actually cool with not including this comment. Basically the code I've added in this commit is needed to not break these kinds of tests, but if someone later removed that code (the code that ignores duplicates that I've added here) and it doesn't break any tests (in the future), thats really not a breaking change (its fundamentally not needed). So perhaps we should just not add this comment!
</blockquote></details>

Seems fine to me (to omit).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: move table Add/Drop/Rename to its own method

remove passing in lease into a defer function.

vivekmenezes commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbRUiLMN6ASibrRkAit:-KbpHnRSZp7D46XA2hOq:b-5ma9jx) (no related file):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

This PR is ready for another review. Thanks!
</blockquote></details>

I've placed back the code that ignores a table descriptor not found error in session.go because that was the easiest way to fix this test without causing much of a headache. This feels like the best way to fix it with adding extra logic all over the place or changing or coordinating things in the test even further. If the application of a schema change returns a descriptor not found error we log a warning and return

---

*[pkg/sql/schema_changer.go, line 283 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-KbkkM5UkqTGJ4VDOFTV:-KbpHi2cbNeLE05n82Vy:b-sqb0o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be401d250922f70f34584d04a406c5ef590c80b/pkg/sql/schema_changer.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Where?
</blockquote></details>

Added a new commit

---

*[pkg/sql/testdata/fk, line 401 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbl6MpbYh03ktRd6hSx:-KbpHlQ6jqik-dniPyEl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/406193b7f30328751d3a29035dcf3c8a08348b29/pkg/sql/testdata/fk#L401)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Seems fine to me (to omit).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166)*
<!-- Sent from Reviewable.io -->

tamird commented:
Still -1 on the first commit; what's the point?

---

Reviewed 4 of 4 files at r8, 3 of 3 files at r9.
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbqGqJpqoE0j1uFoEBA:b-4cspv3)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
It's a distraction from the main execution path that very rarely is added/deleting/renaming a table. Most of the code in exec() calls out to functions that implement specific functionality like MaybeIncrementVersion() and runStateMachineAndBackfill(). It does make exec() easier to read through and know where needRelease is set for instance. 

---

Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbqLs2d3oSV_Px8Khf6:bp037gb)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I think the refactoring in the first commit is fine since that `exec` function was long.

I don't truly understand the 2nd commit though :). What exactly does removal of the `IsDone()` check have to do with refusing to enqueue duplicates? Why are duplicates being enqueued in the first place?
And you say "A duplicate DROP TABLE schema change execution would return
ErrDescriptorNotFound". Is this new behaviour, and, if so, what causes it? What was happening before?

---

Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 441 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbu_UmhDdGt7g4BiQKR:-Kbu_UmhDdGt7g4BiQKS:b-2g80cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b9c2fd9e58df59f56986bf5e244f2c3e515e0a7/pkg/sql/session.go#L441)):*
> ```Go
> 
> func (scc *schemaChangerCollection) queueSchemaChanger(schemaChanger SchemaChanger) {
> 	// Ignore duplicates.
> ```

Why would duplicates be enqueued? Maybe add a comment...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-Kbua-j0upVEYw2Iw2sv:b4ztjyf)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: improve testing of table deletion in the middle of a schema change

vivekmenezes commented:
I've changed the second commit quite a bit and renamed it. I think it's a lot easier to understand. I've also changed the second commit to be the one that actually fixes the issue. TFTR

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/session.go, line 441 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-Kbu_UmhDdGt7g4BiQKR:-KbvQHkToYcpwe3nV67Q:b-b8dhmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b9c2fd9e58df59f56986bf5e244f2c3e515e0a7/pkg/sql/session.go#L441)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Why would duplicates be enqueued? Maybe add a comment...
</blockquote></details>

Removed this code since we handle ErrDescriptorNotFound in the sync path

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-KbvQRoWFxfmJUIHQflA:b-jj8340)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Any more comments on this PR? TFTR

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13166#-:-Kc-GhuHjrYNggbs1w0M:bnfp4nl)*
<!-- Sent from Reviewable.io -->
